plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,169 @@
+    {
+        public TestCase(@Language("SQL") String query, int expectedJdbcConnectionsCount)
+        {
+            this(query, expectedJdbcConnectionsCount, Optional.empty());
+        }
+
+        public TestCase(@Language("SQL") String query, int expectedJdbcConnectionsCount, String errorMessage)
+        {
+            this(query, expectedJdbcConnectionsCount, Optional.of(errorMessage));
+        }
+
+        public TestCase
+        {
+            requireNonNull(query, "query is null");
+            requireNonNull(errorMessage, "errorMessage is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return query;
+        }
+    }
+
+    protected static class ConnectionCountingConnectionFactory
+            implements ConnectionFactory
+    {
+        // Map from connection to a fake exception indicating where the connection was acquired
+        private final Map<Connection, Exception> connectionCreations = synchronizedMap(new IdentityHashMap<>());
+        private final AtomicInteger openConnections = new AtomicInteger();
+        private final AtomicInteger closedConnections = new AtomicInteger();
+
+        private ConnectionFactory delegate;
+
+        public ConnectionCountingConnectionFactory(DriverConnectionFactory delegate)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        public void setDelegate(DriverConnectionFactory delegate)
+        {
+            this.delegate = requireNonNull(delegate, "delegate is null");
+        }
+
+        @Override
+        public Connection openConnection(ConnectorSession session)
+                throws SQLException
+        {
+            openConnections.incrementAndGet();
+            Connection connection = delegate.openConnection(session);
+            connectionCreations.put(connection, new Exception("STACKTRACE"));
+            return new ForwardingConnection()
+            {
+                private volatile boolean closed;
+
+                @Override
+                protected Connection delegate()
+                {
+                    return connection;
+                }
+
+                @Override
+                public void close()
+                        throws SQLException
+                {
+                    if (closed) {
+                        return;
+                    }
+                    closed = true;
+                    connectionCreations.remove(connection);
+                    closedConnections.incrementAndGet();
+                    super.close();
+                }
+            };
+        }
+
+        private void assertThatNoConnectionHasLeaked()
+        {
+            if (!connectionCreations.isEmpty()) {
+                AssertionError error = new AssertionError("%s connections leaked, see attached places".formatted(connectionCreations.size()));
+                connectionCreations.values().forEach(error::addSuppressed);
```

Comment:
should error message note that the suppressed exception is not actually suppressed?

Reply:
it's tests, i don't think it matters.